src/results.rs

```diff
@@ -71,3 +72,94 @@ pub(crate) struct GetMoreResult {
+/// [`Database::list_collections`](../struct.Database.html#method.list_collections).
+pub enum CollectionType {
+    /// Indicates that the data store is a view.
+    View,
+
+    /// Indicates that the data store is a collection.
+    Collection,
+
+    /// An unknown collection type. This is included for forwards compatibility.
+    Other(String),
+}
+
+impl<'de> Deserialize<'de> for CollectionType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de> {
+        let s = String::deserialize(deserializer)?;
+        let out = match s.as_str() {
+            "collection" => CollectionType::Collection,
+            "view" => CollectionType::View,
+            _ => CollectionType::Other(s)
+        };
+        Ok(out)
+    }
+}
+
+/// Info about the collection that is contained in the `CollectionSpecification::info` field of a specification returned from
+/// [`Database::list_collections`](../struct.Database.html#method.list_collections).
+///
+/// See the MongoDB [manual](https://docs.mongodb.com/manual/reference/command/listCollections/#listCollections.cursor)
+/// for more information.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub struct CollectionSpecificationInfo {
+    /// Indicates whether or not the data store is read-only.
+    pub read_only: bool,
+
+    /// The collection's UUID - once established, this does not change and remains the same across replica
+    /// set members and shards in a sharded cluster. If the data store is a view, this field is `None`.
+    pub uuid: Option<Binary>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+/// Information about a collection as reported by [`Database::list_collections`](../struct.Database.html#method.list_collections).
+pub struct CollectionSpecification {
+    /// The name of the collection.
+    pub name: String,
+
+    /// Type of the data store.
+    #[serde(rename="type")]
+    pub collection_type: CollectionType,
+
+    /// The options used to create the collection.
+    pub options: CreateCollectionOptions,
+
+    /// Additional info pertaining to the collection.
+    pub info: CollectionSpecificationInfo,
+
+    /// Provides information on the _id index for the collection
+    /// For views, this is `None`.
+    pub id_index: Option<Document>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+/// A struct modeling the information about an individual database returned from
+/// [`Client::list_databases`](../struct.Client.html#method.list_databases).
+pub struct DatabaseSpecification {
+    /// The name of the database.
+    pub name: String,
+
+    /// The amount of disk space in bytes that is consumed by the database.
+    #[serde(deserialize_with = "crate::bson_util::deserialize_i64_from_bson_number")]
+    pub size_on_disk: i64,
```

Comment:
This value is returned as a bunch of different types depending on the server version, so we need a helper to deserialize it from whatever type to an integer. See [here](https://jira.mongodb.org/browse/SERVER-52544?focusedCommentId=3505206&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-3505206) for an overview if interested.

Reply:
I agree u64 does seem to make sense for a size field

Reply:
I thought so too originally, but this is a value that's deserialized from server responses, meaning it can _possibly_ have negative values. I checked elsewhere in the driver for precedent and saw that we use `i64` for similar values, e.g. [`matched_count`](https://github.com/mongodb/mongo-rust-driver/blob/master/src/results.rs#L52) in `UpdateResult`. That being said, it seems like we probably would be totally fine using an unsigned value here if we wanted to, so I've updated it.

I also updated `get_int` to be a little more lenient when reading integers from doubles in case of a floating point rounding issue.

Reply:
I filed RUST-764 to discuss updating other such fields.